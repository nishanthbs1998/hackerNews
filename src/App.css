.page{
  margin:20px;
}

.interactions{
  text-align: center;
}

.table{
  margin:20px
}
.table-header {
  display: flex;
  line-height: 24px;
  font-size: 16px;
  padding: 0 10px;
  justify-content: space-between;
  }
  .table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
  }

  .table-row {
    display: flex;
    line-height: 24px;
    white-space: nowrap;
    margin: 10px 0;
    padding: 10px;
    background: #ffffff;
    border: 1px solid #e3e3e3;
    }
    .table-header > span {
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 0 5px;
    }
    .table-row > span {
      overflow: hidden;
      text-overflow: ellipsis;
      padding: 0 5px;
      }
    .button-inline {
      border-width: 0;
     /* background: rgb(42, 157, 202);*/
     background:rgb(211, 31, 31);
      color:seashell;
      text-align: inherit;
      -webkit-font-smoothing: inherit;
      padding: 10;
      font-size: inherit;
      cursor: pointer;
    }
    .button-active {
      border-radius: 0;
      border-bottom: 1.5px solid rgb(211, 31, 31);
      }
    .button:hover{
      color:#38BB6C
    }
    .sub{
      background: rgb(42, 157, 202);
      color: seashell; 
    }
    .More{
      background:rgb(42, 157, 202) ;
      color: seashell;
   
      font-size: medium;
      padding:10px
    }
    .button-table-header{
     /* background:rgb(111, 190, 32);*/
     background:inherit;
      color: black;
      padding:10px;
      position:relative;
    }
    .asc::after{
      content: "▲";
      font-weight: 100;
      color:rgb(211, 31, 31);
    }
    .dsc::after{
      content:"▼";
      font-weight: 100;
      color: rgb(211, 31, 31);
    }
    