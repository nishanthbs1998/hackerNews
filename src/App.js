import React from 'react';
import './App.css';
import { faSpinner } from '@fortawesome/free-solid-svg-icons'; 
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {sortBy} from 'lodash';
const DEFAULT_QUERY='redux';
const PATH_BASE='https://hn.algolia.com/api/v1';
const PATH_SEARCH='/search';
const PARAM='query=';
const DEFAULT_PAGE=0;
const PARAM_PAGE="page="
const DEFAULT_HPP=10;
const PARAM_HPP="hitsPerPage="

const SORTS={
   NONE:(list)=>list,
   TITLE:(list)=>sortBy(list,'title'),
   AUTHOR:(list)=>sortBy(list,'author'),
   COMMENTS:(list)=>sortBy(list,'num_comments').reverse(),
   POINTS:(list)=>sortBy(list,'points').reverse()
}

class App extends React.Component{
  
  constructor(props){
    super(props);
    this.state={
     results:null,
     searchKey:'',
    searchTerm:DEFAULT_QUERY,
    isLoading:false,
    sortKey:'NONE',
    isSortReverse:false
  };

    this.needsToSearchTopStories=this.needsToSearchTopStories.bind(this);
    this.fetchTopStories=this.fetchTopStories.bind(this);
    this.setSearchTopStories=this.setSearchTopStories.bind(this);
    this.onDismiss=this.onDismiss.bind(this);
    this.onSearchChange=this.onSearchChange.bind(this);
    this.onSearchSubmit=this.onSearchSubmit.bind(this);
    this.onSort=this.onSort.bind(this);
   }

   onSort(sortKey){
    const isSortReverse=this.state.sortKey===sortKey&& !this.state.isSortReverse;
     this.setState({sortKey,isSortReverse});
   }
   needsToSearchTopStories(searchTerm){
     return !this.state.results[searchTerm]//this way of using objects is called computed properties
   }
   
  setSearchTopStories(result){
    const {hits,page}=result;
//    this.setState(prevState=>{
    const{searchKey,results}=this.state;
    const oldHits=results&&results[searchKey]?results[searchKey].hits:[];
    const updatedHits=[...oldHits,...hits]
    this.setState({isLoading:false,results:{...results,[searchKey]:{hits:updatedHits,page}}});
  //  })
  }

   fetchTopStories(searchTerm,page){
   this.setState({isLoading:true});
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
    .then(response=>response.json())
    .then(result=>this.setSearchTopStories(result));
    
  }
  

  onDismiss(id){
      const {searchKey,results}=this.state;
      const {hits,page}=results[searchKey];

      const isNotId=item=>item.objectID!==id;
      const updatedHits=hits.filter(isNotId);
      this.setState({results:{...results,[searchKey]:{hits:updatedHits,page}}});
  }

  onSearchChange(event){
    this.setState({searchTerm:event.target.value});//event.target means the component that caused the action.Here, its searchTerm
  }

  onSearchSubmit(event){
    const {searchTerm}=this.state;
    this.setState({searchKey:searchTerm});

    if(this.needsToSearchTopStories(searchTerm)){

      this.fetchTopStories(searchTerm,DEFAULT_PAGE);
    }
    event.preventDefault();//To prevent default behaviour of the browser to reload after hitting submit
  }
  
  componentDidMount(){
    console.log("MOUNTED");
    const {searchTerm}=this.state;
    this.setState({searchKey:searchTerm});
    this.fetchTopStories(searchTerm,DEFAULT_PAGE);
  }

  render(){

    const {results,searchTerm,searchKey,isLoading,sortKey,isSortReverse}=this.state
    const page=(results&&results[searchKey]&&results[searchKey].page)||0;
    const list=(results&&results[searchKey]&&results[searchKey].hits)||[];

    return(

     <div className="page">
       <div className="interactions">
       
       <Search
        value={searchTerm}
        onSubmit={this.onSearchSubmit}
        onChange={this.onSearchChange}
       >Search</Search>
       </div>
      <Table //If result is null, nothing is displayed, else table will be displayed
       list={list}
       onDismiss={this.onDismiss}
       sortKey={sortKey}
       isSortReverse={isSortReverse}
       onSort={this.onSort}/>

      <div style={{textAlign:"center"}}>

        <ButtonWithLoading
          isLoading={isLoading}
          onClick={()=>this.fetchTopStories(searchKey,page+1)}
          className="More"> 
           More        
        </ButtonWithLoading>
      
      </div>

    </div>
    );
  }

};

class Search extends React.Component{// converting this func back to class as there is a need to use a lifecycle method to activate focus
//const Search=({value,onChange,onSubmit,children})=>//we can convert this to a function here since this does not have any state or lifecycle
  
componentDidMount(){
    this.input.focus();
  }
  render(){
    const{value,onChange,onSubmit,children}=this.props;
    let input;
  return(
    <form onSubmit={onSubmit}>
    <input type="text"
      value={value}
      onChange={onChange}
      ref ={(node)=>this.input=node}// By doing this, the web site knows where to focus initially with the intention of guiding a user to get started. Here, the cursor is automatically set to search from.
      />
      <Button type="submit" className="sub" onClick={onSubmit}>{children}</Button>
      
    </form>)
  }
}

const Table=({sortKey,onSort,list,onDismiss,isSortReverse})=>{
const sortedList=SORTS[sortKey](list);
const reverseSortedList=isSortReverse?sortedList.reverse():sortedList;
return(
  <div className="table">
    <div className="table-headers">
      <span style={{width:'40%'}}>
        <Sort sortKey='TITLE'
        activateSortKey={sortKey}
        onSort={onSort}>Title</Sort>
      </span>  <span style={{width:'30%'}}>
        <Sort sortKey='AUTHOR'
        activateSortKey={sortKey}
        onSort={onSort}>Author</Sort>
      </span> <span style={{width:'10%'}}>
        <Sort sortKey='COMMENTS'
        activateSortKey={sortKey}
        onSort={onSort}>Comments</Sort>
      </span>  <span style={{width:'10%'}}>
        <Sort sortKey='POINTS'
        activateSortKey={sortKey}
        onSort={onSort}>Points</Sort>
      </span>
    </div>
    {
      reverseSortedList.map(item=>
        <div key={item.objectID} className="table-row">
          <span style={{width:'40%'}}>
            <a href={item.url}>{item.title}</a>
          </span>
          <span style={{width:'30%'}}>{item.author}</span>
          <span style={{width:'10%'}}>{item.num_comments}</span>
          <span style={{width:'10%'}}>{item.points}</span>
          <span style={{width:'10%'}}>
            <Button onClick={()=>onDismiss(item.objectID)}
            className="button-inline">Dismiss</Button>
          </span>
        </div>
      )
    }
    </div>
)
  }
const Button=({onClick,className='',children})=>{
 
    return(
      <button
      type="button"
    onClick={onClick}
    className={className}>{children}</button>

   
    )
  }

  const Loading=()=>
  <div style={{textAlign:"center"}}>
    <FontAwesomeIcon icon={faSpinner} /> 
  </div>

const withLoading=(Component)=>({isLoading,...rest})=>isLoading?<Loading/>: <Component {...rest}/>

const ButtonWithLoading=withLoading(Button);

const Sort=({onSort,sortKey,children,activateSortKey})=>{
  const sortClass=["button-table-header"];
  if(sortKey===activateSortKey){
    sortClass.push('button-active');
  }
  return(
<Button className={sortClass.join(' ')} onClick={()=>onSort(sortKey)}>{children}</Button>//join class to inherit button-active css property along with the original class css property
  )}
export default App;
export{
  Search,
  Table,
  Button
};
