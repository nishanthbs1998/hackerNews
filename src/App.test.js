import React from 'react';
import App from './App';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {Search,Table,Button} from './App';
import Adapter from 'enzyme-adapter-react-16';
import {shallow,configure} from 'enzyme';

configure({adapter:new Adapter()});


describe('App',()=>{
  it('renders',()=>{
    const div=document.createElement('div');
    ReactDOM.render(<App />,div);
  });

    test('snapshot',()=>{
      const component=renderer.create(<App />);
      let tree=component.toJSON();
      expect(tree).toMatchSnapshot();
    })
});

describe('Search',()=>{
  
  it('renders',()=>{
    const div=document.createElement('div');
    ReactDOM.render(<Search>Search</Search>,div);
  });

  test('snapshot',()=>{
    const component=renderer.create(<Search>Search</Search>);
    let tree=component.toJSON();
    expect(tree).toMatchSnapshot();
  })
})

describe('Button',()=>{
  it('renders',()=>{
    const div=document.createElement('div');
    ReactDOM.render(<Button>Button test</Button>,div);
  });

  test('snapshot',()=>{
    const component=renderer.create(<Button>Button test</Button>);
    let tree=component.toJSON();
    expect(tree).toMatchSnapshot();
  })
})

describe('Table',()=>{
  const props={
    list:[{ title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
    { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' }],sortKey:'TITLE',isSortReverse:false
  };

  it('renders',()=>{
    ReactDOM.render(<Table {...props}/>,document.createElement('div'));
  });

  test('snapshot',()=>{
    const component=renderer.create(<Table {...props}/>);
    let tree=component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items in the list',()=>{
    const element=shallow(<Table {...props}/>);
    expect(element.find('.table-row').length).toBe(2);
  })
})

// test('renders learn react link', () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });
