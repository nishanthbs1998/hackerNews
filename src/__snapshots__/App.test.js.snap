// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        className="sub"
        onClick={[Function]}
        type="button"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
     
    <button
      className="More"
      onClick={[Function]}
      type="button"
    >
      More
    </button>
  </div>
</div>
`;

exports[`Button snapshot 1`] = `
<button
  className=""
  type="button"
>
  Button test
</button>
`;

exports[`Search snapshot 1`] = `
<form>
  <input
    type="text"
  />
  <button
    className="sub"
    type="button"
  >
    Search
  </button>
</form>
`;

exports[`Table snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-headers"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-table-header button-active"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
    </span>
      
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-table-header"
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
    </span>
     
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-table-header"
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
    </span>
      
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-table-header"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
